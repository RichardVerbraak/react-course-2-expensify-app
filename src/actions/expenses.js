import database from '../firebase/firebase'

// With firebase (is now asynchronous)
// component calls action generator
// action generator returns function
// component dispatches function (?)
// function runs internally by redux and then has the ability to dispatch other actions 

// ADD_EXPENSE
export const addExpense = (expense) => {
    return {
        type: 'ADD_EXPENSE',
        expense: expense
    }
}

// HARDER THEN IT LOOKS
// This dispatches another dispatch call AFTER it's been stored in firebase upon resolve
// We get an expense passed from the component to this action
// If there isn't one we set it default to an empty object
// Then we start off by creating a function that has access to dispatch (now possible too thunk for creating functions inside of actions)
// If there is one with empty fields we set the expenseData to default values
// We setup to destructure things from the expense we got from the component, and store in firebase AND redux store
// We push on the expense and save it to firebase and THEN, if resolved (promise/asynchronous) we dispatch the action to change the redux store
// .then has access to the reference (because it's now stored in firebase) and we store the firebase id in the redux store as well as the rest
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        // Default values if fields were left empty
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData

        const expense = { description, note, amount, createdAt }

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => {
    return {
        type: 'REMOVE_EXPENSE',
        id: id
    }
}

// Gets the id passed in from component --> removes it in the database --> then upon resolve, dispatches the removeExpense action to remove it from redux store
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({ id }))
        })
    }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => {
    return {
        type: 'EDIT_EXPENSE',
        id: id,
        updates: updates
    }
}

// Returns a function that dispatches editExpense only when firebase has stored the data
export const startEditExpense = (id, updates) => {
    return (dispatch) => {        
        return database.ref(`expenses/${id}`).update(updates)
        .then(() => {
            dispatch(editExpense(id, updates))
        })
    }
}

// SET_EXPENSES
export const setExpenses = (expenses) => {
    return {
        type: 'SET_EXPENSES',
        expenses
    }
}

// I previously couldn't see the new expenses loading on the page because createdAt was zero aka 1 Jan 1970
// I added another .then call after we pushed onto the expenses with forEach but its unecessary, it did still work
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value')
        .then((snapshot) => {  
            const expenses = []                           
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(expenses))
        })       
    }
}