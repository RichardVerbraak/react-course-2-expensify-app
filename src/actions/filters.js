// SET_TEXT_FILTER -- Default is empty string
export const setTextFilter = (text = '') => {
    return {
        type: 'SET_TEXT',
        text: text
    }
}

// SORT_BY_DATE
export const sortByDate = () => {
    return {
        type: 'SORT_BY_DATE'
    }
}

// SORT_BY_AMOUNT
export const sortByAmount = () => {
    return {
        type: 'SORT_BY_AMOUNT'        
    }
}

// SET_START_DATE
export const setStartDate = (date) => {
    return {
        type: 'SET_START_DATE',
        startDate: date
    }
}

// SET_END_DATE
export const setEndDate = (date) => {
    return {
        type: 'SET_END_DATE',
        endDate: date
    }
}