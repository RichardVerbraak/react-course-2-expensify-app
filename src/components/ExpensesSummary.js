import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import getVisibleExpenses from '../selectors/expenses'
import getExpensesTotal from '../selectors/expenses-total'
import numeral from 'numeral'
import dutchLocale from '../../locales/numeral-locale'

numeral.locale('nl-nl')

//# Note: I refactored this because the summary wouldn't update when I used state the previous time
// This isn't the most clean code but it works
// TODO: BREAK UP THE LONG LINE OF CODE BELOW

// Change the inline variables (variables defined in render) to be outside
// Maybe use numeral beforehand instead of doing this inside of render
export class ExpensesSummary extends React.Component {    
    visibleExpenses = () => {
        return this.props.expenses.length
    }
    expensesSum = () => {
        return getExpensesTotal(this.props.expenses)
    }
    render() {
        return (
            <div className="page-header">
                <div className="content-container">
                    {this.visibleExpenses() > 0 && 
                        <h1 className="page-header__title">
                        Viewing <span>{this.visibleExpenses()}</span> {this.visibleExpenses() === 1 ? `expense` : `expenses`} totalling <span>{numeral(this.expensesSum() / 100).format('$0,0.00')}</span></h1>
                    }
                    <div className="page-header__actions">
                        <Link className="button" to="/create">Add Expense</Link>
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)

// {this.state.expenseCount === 1 ? `expense` : `expenses`}
// {numeral(this.state.expensesTotal / 100).format('$0,0.00')}

