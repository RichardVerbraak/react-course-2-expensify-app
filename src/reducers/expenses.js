const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch(action.type) {
        // Gets state and returns new arr with the old + the new expense
        case 'ADD_EXPENSE':
            return [...state, action.expense]
        // Makes a new array except for the ones that didn't pass the test (so it excludes the one that did match)
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => {
                return expense.id !== action.id
            })
        // Map -> creates new arr with the results of the below expression that is done on every element
        // Overrides the old expense array with updated expense
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense
                }
            })
        default:
            return state
    }
}

export default expensesReducer