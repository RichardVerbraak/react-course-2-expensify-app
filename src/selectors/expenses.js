import moment from 'moment'

// Gets the filtered and sorted expenses
// Also sorts by the oldest created expense first or highest amount (?)
// Filters startDates out, based on if they are made on the same day or before the createdAt
// Expenses array gets filtered on the booleans coming from the ternary operator comparisons
// If there is no startDate we won't filter based of off that so we just use true (so it doesnt get filtered out or w/e)
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        // Is 1000 < 0 ? if not : put that one first (most recent)
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })    
}

export default getVisibleExpenses