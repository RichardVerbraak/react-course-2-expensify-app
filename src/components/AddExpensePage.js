import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";

// onSubmit gets passed down to ExpenseForm
// which is were we actually add an expense
export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>

        <div className="content-container">
          <ExpenseForm onSubmit={this.onSubmit}></ExpenseForm>
        </div>
      </div>
    );
  }
}

// Functions that only render JSX is much handier than classes (more notes on this)
// But defining a new function inside of it like the onSubmit(dispatch) 'in-line' of this JSX function is bad
// It has to re-create the function every time this component rerenders, it's not that much of a performance issue but better to avoid it
// const AddExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//             onSubmit={(expense) => {
//                 props.onSubmit(addExpense(expense))
//                 props.history.push('/')
//             }}
//         ></ExpenseForm>
//     </div>
// )

// Abstract away with dispatch (we basically rename dispatch to addExpense) which is now more accesible via props, for writing tests
const mapDispatchToProps = dispatch => {
  return {
    startAddExpense: expense => dispatch(startAddExpense(expense))
  };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
