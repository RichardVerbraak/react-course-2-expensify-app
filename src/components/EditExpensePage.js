import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense, startRemoveExpense } from '../actions/expenses'
 
// The action generator for removeExpense expects an object
// this.props on class components
export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense)                  
        this.props.history.push('/')
    }
    onClick = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id})
        this.props.history.push('/')
    }
    render() {
        return (
        <div>            
            <ExpenseForm
                expense={this.expense}
                onSubmit={this.onSubmit}
            >
            </ExpenseForm>
            <button onClick={this.onClick}>Remove</button>
        </div>
        )
    }
}

// Give EditExpensePage access to the expenses array
// Give the page only the expense that matches the id in the url parameter and it's id comes from the :id from the router thing
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    }
}

// 'builts in' dispatch inside of the edit and remove expense funcs
// It basically makes a new function that uses dispatch
// Props can get used here but we dont need it, these would be the props getting passed into the connected component
const mapDispatchToProps = (dispatch, props) => {
    return {
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        startRemoveExpense: ({id}) => dispatch(startRemoveExpense({id}))
    }
}


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)