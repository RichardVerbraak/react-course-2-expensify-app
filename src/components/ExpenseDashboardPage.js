import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesSummary from './ExpensesSummary'

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary></ExpensesSummary>
        <ExpenseListFilters></ExpenseListFilters>
        <ExpenseList></ExpenseList>        
    </div>
)

export default ExpenseDashboardPage