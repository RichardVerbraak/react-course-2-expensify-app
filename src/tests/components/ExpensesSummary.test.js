import React from 'react'
import { shallow } from 'enzyme'
import {ExpensesSummary} from '../../components/ExpensesSummary'
import expenses from '../fixtures/expenses'
import {filters} from '../fixtures/filters'

test('Should render ExpensesSummary with one expense', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} filters={filters}/>)
    expect(wrapper).toMatchSnapshot()
})

test('Should render ExpensesSummary with two expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0], expenses[1]]} filters={filters}/>)
    expect(wrapper).toMatchSnapshot()
})