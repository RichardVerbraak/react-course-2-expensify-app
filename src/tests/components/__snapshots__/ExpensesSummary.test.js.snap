// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        € 1,95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should render ExpensesSummary with two expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       totalling 
      <span>
        € 1.096,95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
